File: week-07/bkryusim/코딩_테스트_공부.py
import math
from collections import deque


def solution(alp, cop, problems):
    alp_target = alp
    cop_target = cop
    for problem in problems:
        if problem[0] > alp_target:
            alp_target = problem[0]
        if problem[1] > cop_target:
            cop_target = problem[1]

    if alp_target == alp and cop_target == cop:
        return 0

    queue = deque()

    dp = [[math.inf for _ in range(cop_target + 1)] for _ in range(alp_target + 1)]
    visited = [[False for _ in range(cop_target + 1)] for _ in range(alp_target + 1)]

    dp[alp][cop] = 0
    queue.append((alp, cop))
    visited[alp][cop] = True

    while queue:
        r, c = queue.popleft()

        # (row, column, cost)
        candidates = []

        # 오른쪽, 아래쪽 검사해서 큐에 넣고, cost 1로 이동할 수 있는 후보로 등록
        if r + 1 <= alp_target:
            candidates.append((r + 1, c, 1))
            if not visited[r + 1][c]:
                queue.append((r + 1, c))
                visited[r + 1][c] = True

        if c + 1 <= cop_target:
            candidates.append((r, c + 1, 1))
            if not visited[r][c + 1]:
                queue.append((r, c + 1))
                visited[r][c + 1] = True

        # 문제에 따라 풀 수 있다면 다음 좌표를 후보로 등록
        for problem in problems:
            if r >= problem[0] and c >= problem[1]:
                candidates.append((min(alp_target, r + problem[2]), min(cop_target, c + problem[3]), problem[4]))

        # 후보 위치의 값에 대소 비교 후 값 업데이트
        for candidate in candidates:
            if dp[candidate[0]][candidate[1]] > dp[r][c] + candidate[2]:
                dp[candidate[0]][candidate[1]] = dp[r][c] + candidate[2]

    return dp[alp_target][cop_target]

File: week-7/dahyen0o/코딩-테스트-공부.java
import java.util.*;

class Solution {
    
    private static int ALP = 0;
    private static int COP = 0;
    
    public int solution(final int alp, final int cop, final int[][] problems) {
        // 목표 alp, cop 조사
        for(int idx = 0; idx < problems.length; idx++) {
            ALP = Math.max(ALP, problems[idx][0]);
            COP = Math.max(COP, problems[idx][1]);
        }
        
        final int[][] dp = new int[150 + 1][150 + 1];
        
        // 초기값 지정 (문제를 하나도 풀지 않았을 때)
        for(int a = 0; a < dp.length; a++) {
            for(int c = 0; c < dp[a].length; c++) {
                if(a > alp) {
                    dp[a][c] += a - alp;
                }
                if(c > cop) {
                    dp[a][c] += c - cop;
                }
            }
        }
        
        // 문제를 최소 하나 풀었을 때 값 업데이트
        for(int a = 0; a < dp.length; a++) {
            for(int c = 0; c < dp[a].length; c++) {
                if(dp[a][c] == 0) continue; 
                
                for(final int[] prob : problems) {
                    // 해당 문제를 풀 수 있는지 검사
                    final int beforeA = a - prob[2];
                    final int beforeC = c - prob[3];
                    if(beforeA < 0 || beforeC < 0) continue;
                    if(beforeA < prob[0] || beforeC < prob[1]) continue;
                    
                    for(int aa = beforeA; aa <= a; aa++) {
                        for(int cc = beforeC; cc <= c; cc++) {
                            dp[a][c] = Math.min(dp[a][c], dp[aa][cc] + prob[4]);
                        }
                    }
                    // dp[a][c] = Math.min(dp[a][c], dp[beforeA][beforeC] + prob[4]);
                    // dp[a][c] = Math.min(dp[a][c], dp[beforeA][c] + prob[4]);
                    // dp[a][c] = Math.min(dp[a][c], dp[a][beforeC] + prob[4]);
                }
            }
        }
              
        int answer = dp[ALP][COP];
        // for(int a = ALP; a < dp.length; a++) {
        //     for(int c = COP; c < dp[a].length; c++) {
        //         answer = Math.min(answer, dp[a][c]);
        //     }
        // }

        return answer;
    }
    
    private void print(final int[][] dp) {
        for(int a = 0; a < dp.length; a++) {
            for(int c = 0; c < dp[a].length; c++) {
                System.out.printf("%3d ", dp[a][c]);
            }
            System.out.println();
        }
    }
}
